From 824dd4b58ca631c2c5001bcf45907cea855e58f7 Mon Sep 17 00:00:00 2001
From: liang ding <liang.ding@intel.com>
Date: Fri, 29 Dec 2017 16:38:20 +0800
Subject: [PATCH 0421/1240] VHM: add interrupt injection support

VHM provides interrupt injection service for emulated devices.
this patch added interrupt injection support APIs.

Change-Id: I10385318877aa52026d6d2fc56f5fdbc8106bbd9
Signed-off-by: liang ding <liang.ding@intel.com>
Signed-off-by: Xiao Zheng <xiao.zheng@intel.com>
Signed-off-by: Jason Chen CJ <jason.cj.chen@intel.com>
Signed-off-by: Mingqiang Chi <mingqiang.chi@intel.com>
Reviewed-by: Dong, Eddie <eddie.dong@intel.com>
Tested-by: Dong, Eddie <eddie.dong@intel.com>
---
 drivers/char/vhm/vhm_dev.c         | 29 +++++++++++++++++++
 drivers/vhm/vhm_hypercall.c        | 59 ++++++++++++++++++++++++++++++++++++++
 drivers/vhm/vhm_vm_mngt.c          | 18 ++++++++++++
 include/linux/vhm/cwp_common.h     | 32 +++++++++++++++++++++
 include/linux/vhm/cwp_hv_defs.h    |  7 +++++
 include/linux/vhm/vhm_hypercall.h  |  4 +++
 include/linux/vhm/vhm_ioctl_defs.h |  7 +++++
 include/linux/vhm/vhm_vm_mngt.h    |  2 ++
 8 files changed, 158 insertions(+)

diff --git a/drivers/char/vhm/vhm_dev.c b/drivers/char/vhm/vhm_dev.c
index 367d4f1..590c256 100644
--- a/drivers/char/vhm/vhm_dev.c
+++ b/drivers/char/vhm/vhm_dev.c
@@ -238,6 +238,35 @@ static long vhm_dev_ioctl(struct file *filep,
 		break;
 	}
 
+	case IC_ASSERT_IRQLINE: {
+		ret = vhm_assert_irqline(vm, ioctl_param);
+		break;
+	}
+
+	case IC_DEASSERT_IRQLINE: {
+		ret = vhm_deassert_irqline(vm, ioctl_param);
+		break;
+	}
+
+	case IC_PULSE_IRQLINE: {
+		ret = vhm_pulse_irqline(vm, ioctl_param);
+		break;
+	}
+
+	case IC_INJECT_MSI: {
+		struct cwp_msi_entry msi;
+
+		if (copy_from_user(&msi, (void *)ioctl_param, sizeof(msi)))
+			return -EFAULT;
+
+		ret = hcall_inject_msi(vm->vmid, virt_to_phys(&msi));
+		if (ret < 0) {
+			pr_err("vhm: failed to inject!\n");
+			return -EFAULT;
+		}
+		break;
+	}
+
 	default:
 		pr_warn("Unknown IOCTL 0x%x\n", ioctl_num);
 		ret = 0;
diff --git a/drivers/vhm/vhm_hypercall.c b/drivers/vhm/vhm_hypercall.c
index 4412e29..cf4c76e 100644
--- a/drivers/vhm/vhm_hypercall.c
+++ b/drivers/vhm/vhm_hypercall.c
@@ -53,6 +53,11 @@
 #include <linux/vhm/cwp_hv_defs.h>
 #include <linux/vhm/vhm_hypercall.h>
 
+inline long hcall_inject_msi(unsigned long vmid, unsigned long msi)
+{
+	return cwp_hypercall2(HC_INJECT_MSI, vmid, msi);
+}
+
 inline long hcall_set_ioreq_buffer(unsigned long vmid, unsigned long buffer)
 {
 	return cwp_hypercall2(HC_SET_IOREQ_BUFFER, vmid, buffer);
@@ -147,3 +152,57 @@ inline long vhm_query_vm_state(struct vhm_vm *vm)
 
 	return ret;
 }
+
+inline long vhm_assert_irqline(struct vhm_vm *vm, unsigned long ioctl_param)
+{
+	long ret = 0;
+	struct cwp_irqline irq;
+
+	if (copy_from_user(&irq, (void *)ioctl_param, sizeof(irq)))
+		return -EFAULT;
+
+	ret = cwp_hypercall2(HC_ASSERT_IRQLINE, vm->vmid,
+			virt_to_phys(&irq));
+	if (ret < 0) {
+		pr_err("vhm: failed to assert irq!\n");
+		return -EFAULT;
+	}
+
+	return ret;
+}
+
+inline long vhm_deassert_irqline(struct vhm_vm *vm, unsigned long ioctl_param)
+{
+	long ret = 0;
+	struct cwp_irqline irq;
+
+	if (copy_from_user(&irq, (void *)ioctl_param, sizeof(irq)))
+		return -EFAULT;
+
+	ret = cwp_hypercall2(HC_DEASSERT_IRQLINE, vm->vmid,
+			virt_to_phys(&irq));
+	if (ret < 0) {
+		pr_err("vhm: failed to deassert irq!\n");
+		return -EFAULT;
+	}
+
+	return ret;
+}
+
+inline long vhm_pulse_irqline(struct vhm_vm *vm, unsigned long ioctl_param)
+{
+	long ret = 0;
+	struct cwp_irqline irq;
+
+	if (copy_from_user(&irq, (void *)ioctl_param, sizeof(irq)))
+		return -EFAULT;
+
+	ret = cwp_hypercall2(HC_PULSE_IRQLINE, vm->vmid,
+			virt_to_phys(&irq));
+	if (ret < 0) {
+		pr_err("vhm: failed to assert irq!\n");
+		return -EFAULT;
+	}
+
+	return ret;
+}
diff --git a/drivers/vhm/vhm_vm_mngt.c b/drivers/vhm/vhm_vm_mngt.c
index 29849e9..fbca6c3 100644
--- a/drivers/vhm/vhm_vm_mngt.c
+++ b/drivers/vhm/vhm_vm_mngt.c
@@ -95,6 +95,24 @@ void put_vm(struct vhm_vm *vm)
 	mutex_unlock(&vhm_vm_list_lock);
 }
 
+int vhm_inject_msi(unsigned long vmid, unsigned long msi_addr,
+		unsigned long msi_data)
+{
+	struct cwp_msi_entry msi;
+	int ret;
+
+	/* msi_addr: addr[19:12] with dest vcpu id */
+	/* msi_data: data[7:0] with vector */
+	msi.msi_addr = msi_addr;
+	msi.msi_data = msi_data;
+	ret = hcall_inject_msi(vmid, virt_to_phys(&msi));
+	if (ret < 0) {
+		pr_err("vhm: failed to inject!\n");
+		return -EFAULT;
+	}
+	return 0;
+}
+
 void vm_list_add(struct list_head *list)
 {
 	list_add(list, &vhm_vm_list);
diff --git a/include/linux/vhm/cwp_common.h b/include/linux/vhm/cwp_common.h
index b91e241..0a1bc83 100644
--- a/include/linux/vhm/cwp_common.h
+++ b/include/linux/vhm/cwp_common.h
@@ -56,6 +56,20 @@
  * Commmon structures for CWP/VHM/DM
  */
 
+enum irq_mode {
+	IRQ_PULSE,
+	IRQ_ASSERT,
+	IRQ_DEASSERT,
+} __attribute__((aligned(4)));
+
+/* ISA type
+ * inject interrut to both PIC and IOAPIC
+ */
+enum interrupt_type {
+	CWP_INTR_TYPE_ISA,
+	CWP_INTR_TYPE_IOAPIC,
+} __attribute__((aligned(4)));
+
 /*
  * IO request
  */
@@ -179,4 +193,22 @@ struct cwp_ioreq_notify {
 	unsigned long vcpu_mask;
 } __attribute__((aligned(8)));
 
+/* For ISA, PIC, IOAPIC etc */
+struct cwp_irqline {
+	enum interrupt_type intr_type;
+	unsigned long pic_irq;        /* IN: for ISA type */
+	unsigned long ioapic_irq;    /* IN: for IOAPIC type, -1 don't inject */
+} __attribute__((aligned(8)));
+
+/* For MSI type inject */
+struct cwp_msi_entry {
+	unsigned long msi_addr;	/* IN: addr[19:12] with dest vcpu id */
+	unsigned long msi_data;	/* IN: data[7:0] with vector */
+} __attribute__((aligned(8)));
+
+/* For NMI inject */
+struct cwp_nmi_entry {
+	unsigned long vcpuid;	/* IN: -1 means vcpu0 */
+} __attribute__((aligned(8)));
+
 #endif /* CWP_COMMON_H */
diff --git a/include/linux/vhm/cwp_hv_defs.h b/include/linux/vhm/cwp_hv_defs.h
index 9bf06bc..0688114 100644
--- a/include/linux/vhm/cwp_hv_defs.h
+++ b/include/linux/vhm/cwp_hv_defs.h
@@ -74,6 +74,13 @@
 #define HC_PAUSE_VM                 _HC_ID(HC_ID, HC_ID_VM_BASE + 0x04)
 #define HC_QUERY_VMSTATE            _HC_ID(HC_ID, HC_ID_VM_BASE + 0x05)
 
+/* IRQ and Interrupts */
+#define HC_ID_IRQ_BASE              0x100UL
+#define HC_ASSERT_IRQLINE           _HC_ID(HC_ID, HC_ID_IRQ_BASE + 0x00)
+#define HC_DEASSERT_IRQLINE         _HC_ID(HC_ID, HC_ID_IRQ_BASE + 0x01)
+#define HC_PULSE_IRQLINE            _HC_ID(HC_ID, HC_ID_IRQ_BASE + 0x02)
+#define HC_INJECT_MSI               _HC_ID(HC_ID, HC_ID_IRQ_BASE + 0x03)
+
 /* DM ioreq management */
 #define HC_ID_IOREQ_BASE            0x200UL
 #define HC_SET_IOREQ_BUFFER         _HC_ID(HC_ID, HC_ID_IOREQ_BASE + 0x00)
diff --git a/include/linux/vhm/vhm_hypercall.h b/include/linux/vhm/vhm_hypercall.h
index 043a08d..ad2c977 100644
--- a/include/linux/vhm/vhm_hypercall.h
+++ b/include/linux/vhm/vhm_hypercall.h
@@ -139,6 +139,7 @@ static inline long cwp_hypercall4(unsigned long hyp_id, unsigned long param1,
 	return result;
 }
 
+inline long hcall_inject_msi(unsigned long vmid, unsigned long msi);
 inline long hcall_set_ioreq_buffer(unsigned long vmid, unsigned long buffer);
 inline long hcall_notify_req_finish(unsigned long vmid,
 		unsigned long vcpu_mask);
@@ -148,5 +149,8 @@ inline long vhm_resume_vm(struct vhm_vm *vm);
 inline long vhm_pause_vm(struct vhm_vm *vm);
 inline long vhm_destroy_vm(struct vhm_vm *vm);
 inline long vhm_query_vm_state(struct vhm_vm *vm);
+inline long vhm_assert_irqline(struct vhm_vm *vm, unsigned long ioctl_param);
+inline long vhm_deassert_irqline(struct vhm_vm *vm, unsigned long ioctl_param);
+inline long vhm_pulse_irqline(struct vhm_vm *vm, unsigned long ioctl_param);
 
 #endif /* VHM_HYPERCALL_H */
diff --git a/include/linux/vhm/vhm_ioctl_defs.h b/include/linux/vhm/vhm_ioctl_defs.h
index 7921c1d..e0a6425 100644
--- a/include/linux/vhm/vhm_ioctl_defs.h
+++ b/include/linux/vhm/vhm_ioctl_defs.h
@@ -64,6 +64,13 @@
 #define IC_PAUSE_VM                    _IC_ID(IC_ID, IC_ID_VM_BASE + 0x04)
 #define IC_QUERY_VMSTATE               _IC_ID(IC_ID, IC_ID_VM_BASE + 0x05)
 
+/* IRQ and Interrupts */
+#define IC_ID_IRQ_BASE                 0x100UL
+#define IC_ASSERT_IRQLINE              _IC_ID(IC_ID, IC_ID_IRQ_BASE + 0x00)
+#define IC_DEASSERT_IRQLINE            _IC_ID(IC_ID, IC_ID_IRQ_BASE + 0x01)
+#define IC_PULSE_IRQLINE               _IC_ID(IC_ID, IC_ID_IRQ_BASE + 0x02)
+#define IC_INJECT_MSI                  _IC_ID(IC_ID, IC_ID_IRQ_BASE + 0x03)
+
 /* DM ioreq management */
 #define IC_ID_IOREQ_BASE                0x200UL
 #define IC_SET_IOREQ_BUFFER             _IC_ID(IC_ID, IC_ID_IOREQ_BASE + 0x00)
diff --git a/include/linux/vhm/vhm_vm_mngt.h b/include/linux/vhm/vhm_vm_mngt.h
index eb41002..fb02c00 100644
--- a/include/linux/vhm/vhm_vm_mngt.h
+++ b/include/linux/vhm/vhm_vm_mngt.h
@@ -77,6 +77,8 @@ struct vhm_vm {
 
 struct vhm_vm *find_get_vm(unsigned long vmid);
 void put_vm(struct vhm_vm *vm);
+int vhm_inject_msi(unsigned long vmid, unsigned long msi_addr,
+	unsigned long msi_data);
 
 void vm_list_add(struct list_head *list);
 void vm_mutex_lock(struct mutex *mlock);
-- 
2.7.4

